internal/neofs/epoch.go

```diff
@@ -0,0 +1,71 @@
+package neofs
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"github.com/nspcc-dev/neofs-sdk-go/client"
+	"github.com/nspcc-dev/neofs-sdk-go/netmap"
+	"go.uber.org/zap"
+)
+
+type (
+	// NetworkInfoGetter represents provider to get actual [netmap.NetworkInfo].
+	NetworkInfoGetter interface {
+		NetworkInfo(ctx context.Context, prm client.PrmNetworkInfo) (netmap.NetworkInfo, error)
+	}
+
+	// EpochGetter represents provider to get actual NeoFS epoch.
+	EpochGetter interface {
+		CurrentEpoch() uint64
+	}
+
+	// PeriodicGetter implements [EpochGetter].
+	PeriodicGetter struct {
+		logger    *zap.Logger
+		netGetter NetworkInfoGetter
+		epoch     atomic.Uint64
+		timeOut   time.Duration
+	}
+)
+
+// NewPeriodicGetter is a constructor to [PeriodicGetter].
+func NewPeriodicGetter(ctx context.Context, initialEpoch uint64, timeOut time.Duration, netGetter NetworkInfoGetter, logger *zap.Logger) *PeriodicGetter {
+	getter := &PeriodicGetter{
+		timeOut:   timeOut,
+		netGetter: netGetter,
+		logger:    logger,
+	}
+
+	getter.epoch.Store(initialEpoch)
+
+	go getter.update(ctx)
+
+	return getter
+}
+
+// CurrentEpoch returns actual epoch.
+//
+// CurrentEpoch implements [EpochGetter].
+func (g *PeriodicGetter) CurrentEpoch() uint64 {
+	return g.epoch.Load()
+}
+
+func (g *PeriodicGetter) update(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
```

Comment:
You don't (and in fact can't right now) stop the timer in this case.

Reply:
Updated